### PR TITLE
ENH: add .serialize() & .deserialize support for expressions

### LINKED ISSUE #1267: serialize ibis expressions
xref #1206 

We want to take an ibis expression, and serialize to JSON (or other format), send over the wire, then de-serialize back to an ibis expression.

``expr.to_json()``
``ibis.client.Client.from_json()``

might be a suitable interface.

### PR DESCRIPTION
**Pull request by jreback**
closes #1267

### CODE REVIEW

**Comment by jreback** [comment]
so this is *much* simpler and built in. I think we should actually still make ``.serialize()`` methods and such and just wrap a simple dict around an expression to add a bit more meta-data though, e.g.
```
In [1]: import ibis
In [2]: from ibis.compat import pickle
In [3]: purchases = ibis.table([('region', 'string'),
   ...:                         ('kind', 'string'),
   ...:                         ('user', 'int64'),
   ...:                         ('amount', 'double')], 'purchases')
   ...: metric = purchases.amount.sum().name('total')

In [4]: metric
Out[4]: 
ref_0
UnboundTable[table]
  name: purchases
  schema:
    region : string
    kind : string
    user : int64
    amount : double

total = Sum[double]
  amount = Column[double*] 'amount' from table
    ref_0
  where:
    None

In [5]: d = {'version': ibis.__version__, 'expr': metric}

In [7]: serialized = pickle.dumps(d)

In [8]: d2 = pickle.loads(serialized)
In [9]: d2
Out[9]: 
{'expr': ref_0
 UnboundTable[table]
   name: purchases
   schema:
     region : string
     kind : string
     user : int64
     amount : double
 
 total = Sum[double]
   amount = Column[double*] 'amount' from table
     ref_0
   where:
     None, 'version': 'v0.12.0+25.ge0120af'}
```

File: ibis/expr/serialize.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    Returns
+    -------
+    tuple(serialied exprion, metadata dict)
+    """
+
+    d = {'version': ibis.__version__,
+         'metadata': metadata,
+         'format': 'pickle'}
+    return pickle.dumps(expr), d
```
Comment (cpcloud) [inline ibis/expr/serialize.py]: Why not serialize the dict with an additional field `'expr': expr`?

**Comment by jreback** [comment]
fixed up

**Comment by jreback** [comment]
@cpcloud @kszucs 

something change with clickhouse? these are reported as failing

```
test_null_column - ibis.clickhouse.tests.test_functions
ibis/clickhouse/tests/test_functions.py
moreibis/clickhouse/tests/test_functions.py:541: in test_null_column
test_typeof[value5-Null] - ibis.clickhouse.tests.test_functions
ibis/clickhouse/tests/test_functions.py
moreAssertionError: assert 'Nullable(Nothing)' == 'Null'   - Nullable(Nothing)   + Null
```

**Comment by jreback** [comment]
rebased

**Comment by kszucs** [comment]
Sorry, I've just read. There were changes with the NullType handling.

**Comment by jreback** [comment]
@cpcloud i should rebase this

**Comment by xmnlab** [comment]
hey @jreback, what is the status of this PR?

there are one file with conflict. could you rebase your code?

**Comment by toryhaavik** [comment]
@jreback let's resurrect this and get it in!

**Comment by datapythonista** [comment]
@jreback what do you want to do with this one? From the conversation I think it was almost ready, do you want me to rebase it?

**Comment by jreback** [comment]
this is just a wrapper around pickle but let me look

**Comment by datapythonista** [comment]
Closing this as stale. Feel free to reopen if you still want to work on it.

**Pull request closed without merging**